File: Windows/32bit/Csharp/advanced/histomode_extcontrol/histomode.cs
/************************************************************************

  Demo access to MultiHarp 150/160 hardware via MHLIB v 3.1

  THIS IS AN ADVANCED DEMO. DO NOT USE FOR YOUR FIRST EXPERIMENTS.
  Look at the variable meascontrol down below to see what it does.

  The program performs a measurement based on hardcoded settings.
  The resulting histogram is stored in an ASCII output file.

  Michael Wahl, PicoQuant GmbH, March 2022

  Note: This is a console application

  Note: At the API level channel numbers are indexed 0..N-1 
    where N is the number of channels the device has.

  
  Tested with the following compilers:

  - MS Visual Studio 2013 and 2019
  - Mono 6.12.0 (Windows)
  - Mono 6.8.0 (Linux)

************************************************************************/


using System;  // for Console
using System.Text;  // for StringBuilder 
using System.IO;  // for File





class HistoMode 
{

  static void Main() 
  {

    int i,j;
    int retcode;
    string cmd = "";
    int[] dev= new int[Constants.MAXDEVNUM];
    int found = 0;
    int NumChannels = 0;

    StringBuilder LibVer = new StringBuilder (8);
    StringBuilder Serial = new StringBuilder (8);
    StringBuilder Errstr = new StringBuilder (40);
    StringBuilder Model  = new StringBuilder (32);
    StringBuilder Partno = new StringBuilder (16);
    StringBuilder Version = new StringBuilder(16);
    StringBuilder Wtext  = new StringBuilder (16384);

    int HistLen;
    int Binning = 0;  // you can change this, observe limits
    int Offset = 0;  // you can change this, observe limits
    int Tacq = 100;  // Measurement time in millisec, you can change this, observe limits

    int SyncDivider = 1;  // you can change this, observe limits

    int SyncTrigEdge = 0;  // you can change this, observe limits
    int SyncTrigLevel = -100;  // you can change this, observe limits
    int InputTrigEdge = 0;  // you can change this, observe limits
    int InputTrigLevel = -100;  // you can change this, observe limits

    double Resolution = 0;

    int Syncrate = 0;
    int Countrate = 0;
    double Integralcount;
    double elapsed;
    int ctcstatus = 0;
    int flags = 0;
    int warnings = 0;

    int meascontrol
      = Constants.MEASCTRL_SINGLESHOT_CTC;    // start by software and stop when CTC expires (default)
    // = Constants.MEASCTRL_C1_GATED;           // measure while C1 is active		1
    // = Constants.MEASCTRL_C1_START_CTC_STOP; // start with C1 and stop when CTC expires 
    // = Constants.MEASCTRL_C1_START_C2_STOP;  // start with C1 and stop with C2
    int edge1 = Constants.EDGE_RISING;  //Edge of C1 to start (if applicable in chosen mode)
    int edge2 = Constants.EDGE_FALLING; //Edge of C2 to stop (if applicable in chosen mode)


    uint[][] counts = new uint[Constants.MAXINPCHAN][];
    for (i = 0; i < Constants.MAXINPCHAN; i++)		
    counts[i] = new uint[Constants.MAXHISTLEN];

    StreamWriter SW = null;

    Console.WriteLine ("MultiHarp MHLib Demo Application                       PicoQuant GmbH, 2022");
    Console.WriteLine ("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");

    try
    {
        retcode = mhlib.MH_GetLibraryVersion(LibVer);
    }
    catch(Exception e)
    {
        Console.WriteLine("Error loading MHLib as \"" + mhlib.MHLib + "\" => " + e.Message);
        Console.WriteLine("press RETURN to exit");
        Console.ReadLine();
        return;
    }

    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetLibraryVersion error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }
    Console.WriteLine("MHLib Version is " + LibVer);

    if(LibVer.ToString() != Constants.LIB_VERSION)
    {
      Console.WriteLine("This program requires MHLib v." + Constants.LIB_VERSION);
      goto ex;
    }
   
    try
    {
      SW = File.CreateText("histomode.out");
    }
    catch ( Exception )
    {
      Console.WriteLine("Error creating file");
      goto ex;
    }

    Console.WriteLine("Searching for MultiHarp devices...");
    Console.WriteLine("Devidx     Status");

    for(i = 0; i < Constants.MAXDEVNUM; i++)
    {
      retcode = mhlib.MH_OpenDevice(i, Serial);  
      if(retcode == 0) //Grab any MultiHarp we can open
      {
        Console.WriteLine("  {0}        S/N {1}", i, Serial);
        dev[found] = i; //keep index to devices we want to use
        found++;
      }
      else
      {
        if (retcode == Errorcodes.MH_ERROR_DEVICE_OPEN_FAIL)
        {
          Console.WriteLine("  {0}        no device", i);
        }
        else
        {
          mhlib.MH_GetErrorString(Errstr, retcode);
          Console.WriteLine("  {0}        S/N {1}", i, Errstr);
        }
      }
    }

    //In this demo we will use the first device we find, i.e. dev[0].
    //You can also use multiple devices in parallel.
    //You can also check for specific serial numbers, so that you always know 
    //which physical device you are talking to.

    if(found < 1)
    {
      Console.WriteLine("No device available.");
      goto ex; 
    }

    Console.WriteLine("Using device {0}", dev[0]);
    Console.WriteLine("Binning           : {0}", Binning);
    Console.WriteLine("Offset            : {0}", Offset);
    Console.WriteLine("AcquisitionTime   : {0}", Tacq);
    Console.WriteLine("SyncDivider       : {0}", SyncDivider);

    Console.WriteLine("Initializing the device...");

    retcode = mhlib.MH_Initialize(dev[0], Constants.MODE_HIST, 0);  //Histo mode with internal clock
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_Initialize error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    retcode = mhlib.MH_GetHardwareInfo(dev[0], Model, Partno, Version); //this is only for information
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetHardwareInfo error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }
    else
    {
      Console.WriteLine("Found Model {0} Part no {1} Version {2}", Model, Partno, Version);
    }

    retcode = mhlib.MH_GetNumOfInputChannels(dev[0], ref NumChannels); 
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetNumOfInputChannels error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }
    else
    {
      Console.WriteLine("Device has {0} input channels.",NumChannels);
    }

    retcode = mhlib.MH_SetSyncDiv(dev[0], SyncDivider);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetSyncDiv error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    retcode = mhlib.MH_SetSyncEdgeTrg(dev[0], SyncTrigLevel, SyncTrigEdge);
    if (retcode < 0)
    {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_SetSyncEdgeTrg error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
    }
        
    retcode = mhlib.MH_SetSyncChannelOffset(dev[0], 0);    //in ps, emulate a cable delay
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetSyncChannelOffset error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    for(i = 0; i < NumChannels; i++) // we use the same input settings for all channels
    {
      retcode = mhlib.MH_SetInputEdgeTrg(dev[0], i, InputTrigLevel, InputTrigEdge);
      if (retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_SetInputEdgeTrg error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }

      retcode = mhlib.MH_SetInputChannelOffset(dev[0], i, 0);  //in ps, emulate a cable delay
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_SetInputChannelOffset error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }

      retcode = mhlib.MH_SetInputChannelEnable(dev[0], i, 1);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_SetInputChannelEnable error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }
    }

    retcode = mhlib.MH_SetHistoLen(dev[0], Constants.MAXLENCODE, out HistLen);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetHistoLen error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }
    Console.WriteLine("Histogram length is {0}", HistLen);

    retcode = mhlib.MH_SetBinning(dev[0], Binning);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetBinning error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    retcode = mhlib.MH_SetOffset(dev[0], Offset);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetOffset error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    retcode = mhlib.MH_GetResolution(dev[0], ref Resolution);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetResolution error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    Console.WriteLine("Resolution is {0} ps", Resolution);


    // After Init allow 150 ms for valid  count rate readings
    // Subsequently you get new values after every 100ms
    System.Threading.Thread.Sleep(150);

    retcode = mhlib.MH_GetSyncRate(dev[0], ref Syncrate);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetSyncRate error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }
    Console.WriteLine("Syncrate = {0}/s", Syncrate);

    for(i = 0; i < NumChannels; i++) // for all channels
    {
      retcode = mhlib.MH_GetCountRate(dev[0], i, ref Countrate);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_GetCountRate error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }
      Console.WriteLine("Countrate[{0}] = {1}/s", i, Countrate);
    }

    Console.WriteLine();

    // After getting the count rates you can check for warnings
    retcode = mhlib.MH_GetWarnings(dev[0], ref warnings);
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_GetWarnings error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    if(warnings != 0)
    {
      mhlib.MH_GetWarningsText(dev[0], Wtext, warnings);
      Console.WriteLine("{0}", Wtext);
    }

    retcode = mhlib.MH_SetStopOverflow(dev[0], 0, 10000);  // for example only
    if(retcode < 0)
    {
      mhlib.MH_GetErrorString(Errstr, retcode);
      Console.WriteLine("MH_SetStopOverflow error {0} ({1}). Aborted.", retcode, Errstr);
      goto ex;
    }

    retcode = mhlib.MH_SetMeasControl(dev[0], meascontrol, edge1, edge2);
    if (retcode < 0)
    {
        Console.WriteLine("MH_SetMeasControl error {0}. Aborted.", retcode);
        goto ex;
    }

    while(cmd != "q")
    { 
      mhlib.MH_ClearHistMem(dev[0]);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_ClearHistMem error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }

      Console.WriteLine("press RETURN to start measurement");
      Console.ReadLine();

      retcode = mhlib.MH_GetSyncRate(dev[0], ref Syncrate);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_ClearHistMem error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }
      Console.WriteLine("Syncrate = {0}/s", Syncrate);

      for(i = 0; i < NumChannels; i++) // for all channels
      {
        retcode = mhlib.MH_GetCountRate(dev[0], i, ref Countrate);
        if(retcode<0)
        {
          mhlib.MH_GetErrorString(Errstr, retcode);
          Console.WriteLine("MH_GetCountRate error {0} ({1}). Aborted.", retcode, Errstr);
          goto ex;
        }
        Console.WriteLine("Countrate[{0}] = {1}/s", i, Countrate);
      }

      retcode = mhlib.MH_StartMeas(dev[0], Tacq);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_StartMeas error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }

      #region here the demo differs from 'histomode'

      if (meascontrol != Constants.MEASCTRL_SINGLESHOT_CTC)
      {
          Console.WriteLine("waiting for hardware start on C1...");
          ctcstatus = 1;
          while (ctcstatus == 1)
          {
              retcode = mhlib.MH_CTCStatus(dev[0], ref ctcstatus);
              if (retcode < 0)
              {
                  Console.WriteLine("MH_CTCStatus error {0}. Aborted.", retcode);
                  goto ex;
              }
          }
      }

      if ((meascontrol == Constants.MEASCTRL_SINGLESHOT_CTC) || meascontrol == Constants.MEASCTRL_C1_START_CTC_STOP)
          Console.WriteLine("\nMeasuring for {0} milliseconds...", Tacq);

      if (meascontrol == Constants.MEASCTRL_C1_GATED)
          Console.WriteLine("\nMeasuring, waiting for other C1 edge to stop...");

      if (meascontrol == Constants.MEASCTRL_C1_START_C2_STOP)
          Console.WriteLine("\nMeasuring, waiting for C2 to stop...");


      ctcstatus = 0;

      while (ctcstatus == 0)
      {
          retcode = mhlib.MH_CTCStatus(dev[0], ref ctcstatus);
          if (retcode < 0)
          {
              mhlib.MH_GetErrorString(Errstr, retcode);
              Console.WriteLine("\nMH_CTCStatus error {0} ({1}). Aborted.", retcode, Errstr);
              goto ex;
          }
      }

      retcode = mhlib.MH_StopMeas(dev[0]);
      if (retcode < 0)
      {
          mhlib.MH_GetErrorString(Errstr, retcode);
          Console.WriteLine("\nMH_StopMeas error {0} ({1}). Aborted.", retcode, Errstr);
          goto ex;
      }

      retcode = mhlib.MH_GetElapsedMeasTime(dev[0], out elapsed);
      if (retcode < 0)
      {
          Console.WriteLine("\nTH260_GetElapsedMeasTime error{0}. Aborted.", retcode);
          goto ex;
      }
      Console.WriteLine("\n  Elapsed measurement time was{0} ms", elapsed);


      #endregion

      Console.WriteLine();
      for(i = 0; i < NumChannels; i++)  // for all channels
      {
        retcode = mhlib.MH_GetHistogram(dev[0], counts[i], i);
        if(retcode < 0)
        {
          mhlib.MH_GetErrorString(Errstr, retcode);
          Console.WriteLine("MH_GetHistogram error {0} ({1}). Aborted.", retcode, Errstr);
          goto ex;
        }

        Integralcount = 0;
        for(j = 0; j < HistLen; j++)
        {
          Integralcount+=counts[i][j];
        }

        Console.WriteLine("  Integralcount[{0}] = {1}", i, Integralcount);
      }

      Console.WriteLine();

      retcode = mhlib.MH_GetFlags(dev[0], ref flags);
      if(retcode < 0)
      {
        mhlib.MH_GetErrorString(Errstr, retcode);
        Console.WriteLine("MH_GetFlags error {0} ({1}). Aborted.", retcode, Errstr);
        goto ex;
      }

      if ((flags & Constants.FLAG_OVERFLOW) != 0)
      {
        Console.WriteLine("  Overflow.");
      }

      Console.WriteLine("Enter c to continue or q to quit and save the count data.");
      cmd = Console.ReadLine();

    }  // while

    for(j = 0; j < HistLen; j++)
    {
      for(i = 0; i < NumChannels; i++)
      {
        SW.Write("{0,5} ", counts[i][j]);
      }
      SW.WriteLine();
    }

  ex:
    for(i = 0; i < Constants.MAXDEVNUM; i++)  // no harm to close all
    {
      mhlib.MH_CloseDevice(i);
    }

    try
    {
        SW.Flush();
        SW.Close();
        SW.Dispose();
        SW = null;
    }
    catch(Exception e)
    {
        Console.WriteLine("Error closing the file: " + e);
    }
	
    Console.WriteLine("press RETURN to exit");
    Console.ReadLine();
  }
}

File: Linux/Matlab/histomode.m

% Demo for access to MultiHarp 150/160 hardware via MHLIB v 3.0.
% The program performs a measurement based on hard coded settings.
% The resulting histogram (65536 bins) is stored in an ASCII output file.
%
% Axel Hagen, PicoQuant, May 2020
% Michael Wahl, PicoQuant, March 2021

% Constants from mhdefin.h

REQLIBVER   =     '3.0';    % this is the version this program expects
MAXDEVNUM   =         8;
MAXHISTBINS =     65536;	 % number of histogram channels
MAXLENCODE  =         6;	 % max histogram length is 65536	
MAXBINSTEPS =        24;
MODE_HIST   =         0;
MODE_T2	    =         2;
MODE_T3	    =         3;

FLAG_OVERFLOW = hex2dec('0001');

TRGLVLMIN	  =       -1200;	   % mV
TRGLVLMAX	  =        1200;	   % mV
OFFSETMIN	  =           0;	   % ns
OFFSETMAX	  =   100000000;	   % ns

ACQTMIN		  =           1;	   % ms
ACQTMAX		  =   360000000;	   % ms  (100*60*60*1000ms = 100h)

% Errorcodes from errorcodes.h

MH_ERROR_DEVICE_OPEN_FAIL		 = -1;

% Settings for the measurement, Adapt to your setup!

SyncTrgEdge   =    0;     %  you can change this
SyncTrgLevel  = -100;     %  you can change this
InputTrgEdge  =    0;     %  you can change this
InputTrgLevel = -100;     %  you can change this
SyncDiv       =    1;     %  you can change this
Binning       =    0;     %  you can change this
Tacq          = 1000;     %  you can change this      
    
fprintf('\nMultiHarp 150/160 MHLib Demo Application             PicoQuant 2021\n');

if (~libisloaded('MHlib'))    
    %Attention: The header file name given below is case sensitive and must
    %be spelled exactly the same as the actual name on disk except the file 
    %extension. 
    %Wrong case will apparently do the load successfully but you will not
    %be able to access the library!
    %The alias is used to provide a fixed spelling for any further access via
    %calllib() etc, which is also case sensitive.
    
    %To load the right dll for OS and bitness we use the mexext command.
    OS = mexext;
    if strcmp('mexw32', OS)
        DLL = 'MHLib.dll';                        % Windows 32 bit
    elseif strcmp('mexw64', OS)
        DLL = 'MHLib64.dll';                      % Windows 64 bit
    elseif strcmp('mexa32', OS)
        DLL = '/usr/local/lib/mh150/mhlib.so';    % Linux 32 bit
    elseif strcmp('mexa64', OS)
        DLL = '/usr/local/lib64/mh150/mhlib.so';  % Linux 64 bit
    else
        fprintf('\nNo supported OS\n');
        return;
    end
    loadlibrary(DLL, 'mhlib.h', 'alias', 'MHlib');
else
    fprintf('Note: MHlib was already loaded\n');
end

if (libisloaded('MHlib'))
    fprintf('MHlib opened successfully\n');
    %libfunctionsview MHlib; %use this to test for proper loading
else
    fprintf('Could not open MHlib\n');
    return;
end
    
LibVersion    = blanks(8); %enough length!
LibVersionPtr = libpointer('cstring', LibVersion);

[ret, LibVersion] = calllib('MHlib', 'MH_GetLibraryVersion', LibVersionPtr);
if (ret<0)
    fprintf('Error in GetLibVersion. Aborted.\n');
    err = MH_GETLIBVERSION_ERROR;
else
	fprintf('MHLib version is %s\n', LibVersion);
end

if ~strcmp(LibVersion,REQLIBVER)
    fprintf('This program requires MHLib version %s\n', REQLIBVER);
    return;
end

fid = fopen('histomode.out','w');
if (fid<0)
    fprintf('Cannot open output file\n');
    return;
end

 fprintf(fid,'Binning           : %ld\n',Binning);
 fprintf(fid,'AcquisitionTime   : %ld\n',Tacq);
 fprintf(fid,'SyncDivider       : %ld\n',SyncDiv);
 fprintf(fid,'SyncTrgEdge       : %ld\n',SyncTrgEdge);
 fprintf(fid,'SyncTrgLevel      : %ld\n',SyncTrgLevel);
 fprintf(fid,'InputTrgEdge      : %ld\n',InputTrgEdge);
 fprintf(fid,'InputTrgLevel     : %ld\n',InputTrgLevel);


fprintf('\nSearching for MultiHarp devices...');

dev = [];
found = 0;
Serial     = blanks(8); %enough length!
SerialPtr  = libpointer('cstring', Serial);
ErrorStr   = blanks(40); %enough length!
ErrorPtr   = libpointer('cstring', ErrorStr);

for i=0:MAXDEVNUM-1
    [ret, Serial] = calllib('MHlib', 'MH_OpenDevice', i, SerialPtr);
    if (ret==0)       % Grab any MultiHarp we successfully opened
        fprintf('\n  %1d        S/N %s', i, Serial);
        found = found+1;            
        dev(found)=i; %keep index to devices we may want to use
    else
        if(ret==MH_ERROR_DEVICE_OPEN_FAIL)
            fprintf('\n  %1d        no device', i);
        else 
            [ret, ErrorStr] = calllib('MHlib', 'MH_GetErrorString', ErrorPtr, ret);
            fprintf('\n  %1d        %s', i,ErrorStr);
        end
    end
end
    
% In this demo we will use the first MultiHarp device we found, i.e. dev(1).
% If you have nultiple MultiHarp devices you could also check for a specific 
% serial number, so that you always know which physical device you are talking to.

if (found<1)
	fprintf('\nNo device available. Aborted.\n');
	return; 
end

fprintf('\nUsing device #%1d',dev(1));
fprintf('\nInitializing the device...');

[ret] = calllib('MHlib', 'MH_Initialize', dev(1), MODE_HIST, 0); 
if(ret<0)
	fprintf('\nMH_Initialize error %d. Aborted.\n',ret);
    closedev;
	return;
end 

%this is only for information
Model      = blanks(24); %enough length!
Partno     = blanks(8); %enough length!
Version    = blanks(8); %enough length!
ModelPtr   = libpointer('cstring', Model);
PartnoPtr  = libpointer('cstring', Partno);
VersionPtr = libpointer('cstring', Version);

[ret, Model, Partno] = calllib('MHlib', 'MH_GetHardwareInfo', dev(1), ModelPtr, PartnoPtr, VersionPtr);
if (ret<0)
    fprintf('\nMH_GetHardwareInfo error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nFound model %s part number %s version %s', Model, Partno, Version);             
end

NumInpChannels = int32(0);
NumChPtr = libpointer('int32Ptr', NumInpChannels);
[ret,NumInpChannels] = calllib('MHlib', 'MH_GetNumOfInputChannels', dev(1), NumChPtr); 
if (ret<0)
    fprintf('\nMH_GetNumOfInputChannels error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nDevice has %i input channels.', NumInpChannels);             
end

[ret] = calllib('MHlib', 'MH_SetSyncDiv', dev(1), SyncDiv);
if (ret<0)
    fprintf('\nMH_SetSyncDiv error %1d. Aborted.\n',ret);
    closedev;
    return;
end

[ret] = calllib('MHlib', 'MH_SetSyncEdgeTrg', dev(1), SyncTrgLevel, SyncTrgEdge);
if (ret<0)
    fprintf('\nMH_SyncSetEdgeTrg error %ld. Aborted.\n', ret);
    closedev;
    return;
end

 
[ret] = calllib('MHlib', 'MH_SetSyncChannelOffset', dev(1), 0);
if (ret<0)
   fprintf('\nMH_SetSyncChannelOffset error %ld. Aborted.\n', ret);
   closedev;
   return;
end 

for i=0:NumInpChannels-1 % we use the same input settings for all channels
    [ret] = calllib('MHlib', 'MH_SetInputEdgeTrg', dev(1), i, InputTrgLevel, InputTrgEdge);
    if (ret<0)
        fprintf('\nMH_SetInputEdgeTrg error %ld. Aborted.\n', ret);
        closedev;
        return;
    end   
    [ret] = calllib('MHlib', 'MH_SetInputChannelOffset', dev(1), i, 0);
    if (ret<0)
        fprintf('\nMH_SetInputChannelOffset error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
end

HistLen = int32(0);
HistLenPtr = libpointer('int32Ptr', HistLen);
[ret, HistLen] = calllib('MHlib', 'MH_SetHistoLen', dev(1), MAXLENCODE, HistLenPtr);
if (ret<0)
    fprintf('\nMH_SetHistoLen error %ld. Aborted.\n', ret);
    closedev;
    return;
end

[ret] = calllib('MHlib', 'MH_SetBinning', dev(1), Binning);
if (ret<0)
    fprintf('\nMH_SetBinning error %ld. Aborted.\n', ret);
    closedev;
    return;
end
   
[ret] = calllib('MHlib', 'MH_SetOffset', dev(1), 0);
if (ret<0)
    fprintf('\nMH_SetOffset error %ld. Aborted.\n', ret);
    closedev;
    return;
end

ret = calllib('MHlib', 'MH_SetStopOverflow', dev(1), 0, 10000); %for example only 
if (ret<0)
    fprintf('\nMH_SetStopOverflow error %ld. Aborted.\n', ret);
    closedev;
    return;
end
 
Resolution = 0;
ResolutionPtr = libpointer('doublePtr', Resolution);
[ret, Resolution] = calllib('MHlib', 'MH_GetResolution', dev(1), ResolutionPtr);
if (ret<0)
    fprintf('\nMH_GetResolution error %ld. Aborted.\n', ret);
    closedev;
    return;
end
fprintf('\nResolution=%1dps', Resolution);


pause(0.4); % after Init or SetSyncDiv you must allow 400 ms for valid new count rates
            % otherwise you get new values every 100 ms

% From here you can repeat the measurement (with the same settings)


Syncrate = 0;
SyncratePtr = libpointer('int32Ptr', Syncrate);
[ret, Syncrate] = calllib('MHlib', 'MH_GetSyncRate', dev(1), SyncratePtr);
if (ret<0)
    fprintf('\nMH_GetSyncRate error %ld. Aborted.\n', ret);
    closedev;
    return;
end
fprintf('\nSyncrate=%1d/s', Syncrate);
 
for i=0:NumInpChannels-1
	Countrate = 0;
	CountratePtr = libpointer('int32Ptr', Countrate);
	[ret, Countrate] = calllib('MHlib', 'MH_GetCountRate', dev(1), i, CountratePtr);
	if (ret<0)
        fprintf('\nMH_GetCountRate error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
	fprintf('\nCountrate%1d=%1d/s ', i, Countrate);
end

Warnings = 0;
WarningsPtr = libpointer('int32Ptr', Warnings);
[ret, Warnings] = calllib('MHlib', 'MH_GetWarnings', dev(1), WarningsPtr);
if (ret<0)
    fprintf('\nMH_GetWarnings error %ld. Aborted.\n',ret);
    closedev;
    return;
end
if (Warnings~=0)
    Warningstext = blanks(16384); %enough length!
    WtextPtr     = libpointer('cstring', Warningstext);
    [~, Warningstext] = calllib('MHlib', 'MH_GetWarningsText', dev(1), WtextPtr, Warnings);
    fprintf('\n\n%s',Warningstext);
end


ret = calllib('MHlib', 'MH_ClearHistMem', dev(1));    
if (ret<0)
    fprintf('\nMH_ClearHistMem error %ld. Aborted.\n', ret);
    closedev;
    return;
end
        
ret = calllib('MHlib', 'MH_StartMeas', dev(1),Tacq); 
if (ret<0)
    fprintf('\nMH_StartMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end
         
fprintf('\nMeasuring for %1d milliseconds...',Tacq);
        
ctcdone = int32(0);
ctcdonePtr = libpointer('int32Ptr', ctcdone);
while (ctcdone==0)
    [~,ctcdone] = calllib('MHlib', 'MH_CTCStatus', dev(1), ctcdonePtr);
end    
         
ret = calllib('MHlib', 'MH_StopMeas', dev(1)); 
if (ret<0)
    fprintf('\nMH_StopMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end
        
countsbuffer  = uint32(zeros(NumInpChannels,MAXHISTBINS));
for i=0:NumInpChannels-1  
    bufferptr = libpointer('uint32Ptr', countsbuffer(i+1,:));
    [ret,countsbuffer(i+1,:)] = calllib('MHlib', 'MH_GetHistogram', dev(1), bufferptr, i); 
    if (ret<0)
        fprintf('\nMH_GetHistogram error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
end

flags = int32(0);
flagsPtr = libpointer('int32Ptr', flags);
[ret,flags] = calllib('MHlib', 'MH_GetFlags', dev(1), flagsPtr);
if (ret<0)
    fprintf('\nMH_GetFlags error %ld. Aborted.\n', ret);
    closedev;
    return;
end
        
if(bitand(uint32(flags),FLAG_OVERFLOW)) 
    fprintf('  Overflow.');
end    

Integralcount = sum(countsbuffer');        
fprintf('\nTotalCount=%1d', Integralcount);

fprintf('\nSaving file...');

for i=1:MAXHISTBINS
    fprintf(fid,'%7d ', countsbuffer(:,i));
    fprintf(fid,'\n');
end    

plot(countsbuffer')

fprintf('\nData is in histomode.out ');

closedev;
    
if(fid>0) 
    fclose(fid);
end

